### PR TITLE
Adds support for accuracyThreshold

### PR DESCRIPTION
**Pull request by oliverroick**
### Proposed changes in this pull request

Implements #597: Adds support for XLSForm's `accuracyThreshold` to set the minimum GPS accuracy required when collecting coordinates in GeoODK.

- Adds `questionnaires.validators.validate_accuracy`, which checks if a given value is a positive float. This is used in various places to validate input for `accuracyThreshold`.
- Refactor `questionnaires.models.Question.TYPE_CHOICES` into `questionnaires.choices.QUESTION_TYPES` to resolve a circular import between `models` and `validators`
- Adds the new field `gps_accuracy` to `questionnaires.models.Question`.
- Extends `questionnaires.managers.QuestionManager.create_from_dict` so it reads and stores `accuracyThreshold` from XLSForms.
- Extends `xforms.renderers.XFormRenderer` so `accuracyThreshold` is added to rendered XForms, which are used by GeoODK.
- Adds field `gps_accuracy` to `questionnaires.serializers.QuestionSerializer` so the value is read and written via the JSON API.
- Extends `questionnaires.validators.QUESTION_SCHEMA` so `gps_accuracy` is validated using `questionnaires.validators.validate_accuracy`. `questionnaires.validators.validate_schema` is also extended, so fields can be validated using validation functions.

### When should this PR be merged

Should go into the sprint 20 release. 

### Risks

None. 

### Follow-up actions

The rendering of `accuracyThreshold` into XForms needs to be verified using GeoODK. 

#### General

- **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description. 
  - [ ] Review 1
  - [ ] Review 2
- **Is the PR labeled correctly?** It should have the `migration` label if a new migration is added. 
  - [ ] Review 1
  - [ ] Review 2
- **Is the risk level assessment sufficient?** The risks section should contain all risks that might be introduced with the PR and which actions we need to take to mitigate these risks. Possible risks are database migrations, new libraries that need to be installed or changes to deployment scripts.
  - [ ] Review 1
  - [ ] Review 2

#### Functionality

- **Are all requirements met?** Compare implemented functionality with the requirements specification.
  - [ ] Review 1
  - [ ] Review 2
- **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled. 
  - [ ] Review 1
  - [ ] Review 2

#### Code

- **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
  - [ ] Review 1
  - [ ] Review 2
- **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests. 
  - [ ] Review 1
  - [ ] Review 2
- **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/). 
  - [ ] Review 1
  - [ ] Review 2
- **Is the code documented sufficiently?** Large and complex classes, functions or methods must be annotated with comments following our [code-style guidelines](https://devwiki.corp.cadasta.org/Contributing%20Style%20Guide#documentation-and-comments).
  - [ ] Review 1
  - [ ] Review 2
- **Has the scalability of this change been evaluated?**
  - [ ] Review 1
  - [ ] Review 2
- **Is there a maintenance plan in place?**
  - [ ] Review 1
  - [ ] Review 2

#### Tests

- **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components. 
  - [ ] Review 1
  - [ ] Review 2
- **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
  - [ ] Review 1
  - [ ] Review 2
- **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?
  - [ ] Review 1
  - [ ] Review 2

#### Security

- **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
  - [ ] Review 1
  - [ ] Review 2
- **Are all UI and API inputs run through forms or serializers?** 
  - [ ] Review 1
  - [ ] Review 2
- **Are all external inputs validated and sanitized appropriately?**
  - [ ] Review 1
  - [ ] Review 2
- **Does all branching logic have a default case?**
  - [ ] Review 1
  - [ ] Review 2
- **Does this solution handle outliers and edge cases gracefully?**
  - [ ] Review 1
  - [ ] Review 2
- **Are all external communications secured and restricted to SSL?**
  - [ ] Review 1
  - [ ] Review 2

#### Documentation

- **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes must be documented in the [Cadasta Platform Documentation](https://github.com/Cadasta/cadasta-docs).
  - [ ] Review 1
  - [ ] Review 2
- **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented in the [API docs](https://github.com/Cadasta/api-docs).
  - [ ] Review 1
  - [ ] Review 2
- **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki. 
  - [ ] Review 1
  - [ ] Review 2

### CODE REVIEW

File: cadasta/questionnaires/tests/test_validators.py [hunk 1]
```diff
@@ -45,6 +63,10 @@ def test_validate_id_string_contains_whitespace(self):
             "'id_string' cannot be blank or contain whitespace.")
 
 
+def positive(val):
+    return val > 1
```
Comment (seav) [inline cadasta/questionnaires/tests/test_validators.py]: The function name doesn't seem to match the condition. Should this be `val > 0`? (But this is admittedly inconsequential in the grand scheme of things.)

Comment (oliverroick) [inline cadasta/questionnaires/tests/test_validators.py]: Changed that.

**Comment by oliverroick** [comment]
> Does pyxform validate that accuracyThreshold is only present on geo fields? I don't think it makes sense for us to reject XLSForms that have invalid values for accuracyThreshold on non-geo fields. If pyxform doesn't support this validation, should we build it ourselves?

pyxform does not check if `accuracyThreshold` is only present on geo fields. I agree that it probably makes sense to only validate those for geometry fields. I'll work on it tomorrow. 

> For the manual test cases, I suppose we should add test cases for attempting to upload XLSForms with valid and various invalid values (like -1.5, 0, and STRING).

Sounds about right.

**Comment by oliverroick** [comment]
@seav I added code to make sure `accuracyThreshold` is only validated on geometry fields and ignored for all other fields.

File: cadasta/questionnaires/tests/test_managers.py [hunk 2]
```diff
@@ -378,8 +378,32 @@ def test_create_from_dict(self):
...
+
+        assert model.question_group is None
+        assert model.questionnaire == questionnaire
+        assert model.label == question_dict['label']
+        assert model.name == question_dict['name']
+        assert model.type == 'GP'
+        assert model.gps_accuracy == 1.5
+
+    def test_create_from_dict_ingnore_accuracy_threshold(self):
+        """For non-geomtrey fields accuracy should be ignored"""
```
Comment (seav) [inline cadasta/questionnaires/tests/test_managers.py]: **[MINOR]** "geomtrey" should be ""geometry".

File: cadasta/questionnaires/tests/test_validators.py [hunk 3]
```diff
@@ -21,6 +21,12 @@ def test_wrong_type(self):
         assert validators.validate_accuracy('Something') is False
 
 
+class GpsRelevvantTest(TestCase):
```
Comment (seav) [inline cadasta/questionnaires/tests/test_validators.py]: **[MINOR]** `Relevvant` should be `Relevant`.

**Comment by seav** [comment]
(Weird. I check items on the review checklist but they aren't persistent...)

File: cadasta/questionnaires/tests/test_managers.py (hunk 2 above)
Comment (oliverroick) [inline cadasta/questionnaires/tests/test_managers.py]: Fixed

File: cadasta/questionnaires/tests/test_validators.py (hunk 3 above)
Comment (oliverroick) [inline cadasta/questionnaires/tests/test_validators.py]: Fixed

**Comment by oliverroick** [comment]
> (Weird. I check items on the review checklist but they aren't persistent...)

It seems to only work for bugs for some reason...

**Comment by alukach** [comment]
Sorry about the delay, will attempt to get to this over the weekend.

File: cadasta/questionnaires/validators.py [hunk 4]
```diff
@@ -3,7 +3,48 @@
...
+        pass
+
+    return False
+
+
+def gps_relevant(json):
+    return json.get('type') in XFORM_GEOM_FIELDS
+
+
+def validate_id_string(json):
```
Comment (alukach) [inline cadasta/questionnaires/validators.py]: nitpick: `json` doesn't feel like an appropriate name for this field as JSON is a `str` and these are obviously `dict` instances.  Maybe just `data`? 🤷‍♂️

Comment (alukach) [inline cadasta/questionnaires/validators.py]: I see that this is done elsewhere. Maybe it's our convention and we should leave as is.

**Pull request merged**